src/v/kafka/client/partitioners.cc

```diff
@@ -108,4 +110,45 @@ partitioner default_partitioner(model::partition_id initial) {
       detail::roundrobin_partitioner{initial})};
 }
 
+void partitioners_cache::apply_metadata(const metadata_response_data& data) {
+    chunked_hash_set<model::topic> metadata_topics;
+    for (const auto& t : data.topics) {
+        auto it = _partitioners.find(t.name);
+        if (
+          it != _partitioners.end()
+          && it->second.partition_count == t.partitions.size()) {
+            // If the topic already exists with the same partition count,
+            // we can skip it.
+            continue;
+        }
+
+        const auto initial_partition_id = model::partition_id{
+          random_generators::get_int<model::partition_id::type>(
+            t.partitions.size())};
+
+        _partitioners[t.name] = entry{
+          .partition_count = t.partitions.size(),
+          .partitioner = default_partitioner(initial_partition_id)};
+        metadata_topics.emplace(t.name);
+    }
+    // remove partitioners for topics that are no longer in the metadata
+    // response
+    std::erase_if(_partitioners, [&metadata_topics](const auto& entry) {
+        return !metadata_topics.contains(entry.first);
+    });
```

Comment:
shouldn't this look for topics in metadata_response_data because some topics may be skipped because of this condition

```
       && it->second.partition_count == t.partitions.size()) {
            // If the topic already exists with the same partition count,
            // we can skip it.
```

Reply:
good find

Reply:
good find, done